pkg/metrics/metrics.go

```diff
@@ -0,0 +1,179 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package metrics
+
+import (
+	"fmt"
+	"net"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	"k8s.io/apimachinery/pkg/types"
+	k8smetrics "k8s.io/component-base/metrics"
+	"k8s.io/klog"
+)
+
+const (
+	labelDriverName      = "driver_name"
+	labelOperationName   = "operation_name"
+	labelOperationStatus = "operation_status"
+	subSystem            = "snapshot_controller"
+	metricName           = "operation_total_seconds"
+	metricHelpMsg        = "Total number of seconds spent by the controller on an operation from end to end"
+)
+
+type OperationState string
+
+const (
+	// Failed is an operation state which means the controller has encountered
+	// an error which is not recoverable and the controller has marked a permanent
+	// failure of the operation.
+	Failure OperationState = "Failure"
+
+	// Success states that the controller has successfully executed the operation.
+	Success OperationState = "Success"
+)
+
+var metricBuckets = []float64{0.1, 0.25, 0.5, 1, 2.5, 5, 10, 15, 30, 60, 120, 300, 600}
+
+type MetricsManager interface {
+	// StartServing starts the metrics endpoint at the specified addr/pattern for
+	// metrics managed by this MetricsManager. It spawns a goroutine to listen to
+	// and serve HTTP requests received on addr/pattern
+	// if the "pattern" is empty (i.e., ""), no endpoint will be started. An error
+	// will be returned if there is any.
+	StartServing(pattern, addr string, logger promhttp.Logger, wg *sync.WaitGroup) (*http.Server, error)
+
+	// OperationStart takes in an operation and caches its start time.
+	// if the operation already exists, it's an no-op.
+	OperationStart(op Operation)
+
+	// DropOperation removes an operation from cache.
+	// if the operation does not exist, it's an no-op.
+	DropOperation(op Operation)
+
+	// RecordMetrics records a metric point. Note that it will be an no-op if an
+	// operation has not been marked "Started" previously via invoking "OperationStart".
+	// Invoking of RecordMetrics effectively removes the cached entry.
+	// op - the operation which the metric is associated with.
+	// state - the operation state
+	//   Failure: record a failure metric
+	//   Success: record a success metric
+	RecordMetrics(op Operation, state OperationState)
```

Comment:
For failure, I think it would be good to support status_codes, so that we can distinguish between configuration errors, and system errors.

Reply:
good point but that would require major changes in the controller code to be able to generate those return codes.

Reply:
Can we make the metrics future proof by having a status_code dimension instead? We did this for in-tree too, with the expectation that in the future we may add more status codes. See https://github.com/kubernetes/kubernetes/pull/75750/files

Reply:
the state code is an alias to string which itself is already future proof? are you asking to make the function to take in string as "state" type?

Reply:
I was thinking if in the future we decide to support error codes, then having the input be an error code would be more flexible. Especially if we want to take this metrics library and reuse it for other controllers (ie put it in csi-lib-utils).

But if we only want to contain this for snapshot controller for now, then at least I think we should be more descriptive in the current failure string to leave room for future error types.

Reply:
introduced a status interface to make it more extendable.